## Panduan Penggunaan Model

Gunakanlah code di bawah ini untuk melakukan analisis terhadap proses preprocessing yang sudah dilakukan.
Keterangan:
1. Ubahlah dataset (csv file) sesuai dengan data kalian.
2. Gunakanlah dataset sebelum dan sesudah dilakukan proses preprocessing
3. Catat hasilnya
4. Bandingkan kedua hasil tersebut lalu buatlah kesimpulannya.

In [51]:
import numpy as np
import pandas as pd
import seaborn as sb
import os

##### Ubahlah dataset.csv sesuai dengan nama dataset kalian

In [52]:
dataframe = pd.read_csv('glcm_result.csv')
dataframe.head()

Unnamed: 0                                              path   contrast_0  \
0           0                        datasets/ayam\download.jpg  1952.343750   
1           1  datasets/ayam\egg-white-food-protein-162712.jpeg  2828.906250   
2           2                      datasets/ayam\images (1).jpg  2346.796875   
3           3                          datasets/ayam\images.jpg  1988.203125   
4           4             datasets/ayam\IMG_20220818_154051.jpg  3882.773438   

   dissimilarity_0  homogeneity_0  energy_0  correlation_0     entropy_0  \
0         7.656250       0.970093  0.866311       0.712865 -5.105249e+06   
1        11.093750       0.956665  0.816831       0.692669 -7.070763e+06   
2         9.203125       0.964050  0.822255       0.747541 -7.147066e+06   
3         7.796875       0.969543  0.866829       0.704665 -5.051554e+06   
4        15.226562       0.940521  0.703014       0.752067 -1.291644e+07   

      asm_0  contrast_45  ...    entropy_95    asm_90  contrast_135  \
0  0.000013       3084.0  ... -5.105249e+06  0.000013        3044.0   
1  0.000012       4408.0  ... -7.070763e+06  0.000013        4596.0   
2  0.000013       3870.0  ... -7.147066e+06  0.000012        4188.0   
3  0.000013       3256.0  ... -5.051554e+06  0.000013        3220.0   
4  0.000011       8468.0  ... -1.291644e+07  0.000010        8506.0   

   dissimilarity_135  homogeneity_135  energy_135  correlation_135  \
0          11.937255         0.953370    0.850427         0.553960   
1          18.023529         0.929596    0.792249         0.502481   
2          16.423529         0.935846    0.796272         0.551083   
3          12.627451         0.950674    0.849020         0.523450   
4          33.356863         0.869700    0.644671         0.458630   

    entropy_135   asm_135  labels  
0 -5.105249e+06  0.000013    ayam  
1 -7.070763e+06  0.000012    ayam  
2 -7.147066e+06  0.000012    ayam  
3 -5.051554e+06  0.000013    ayam  
4 -1.291644e+07  0.000010    ayam  

[5 rows x 31 columns]

In [53]:
# os.listdir(
#     'segmented-dataset'
# )

##### Ubahlah label sesuai dengan label kalian

In [54]:
clean_data  = dataframe.drop(columns=['Unnamed: 0', 'path'])
clean_data  = clean_data.replace ({
    'ayam': 0,
    'kalkun': 1,
    'puyuh': 2
})
clean_data.head()

contrast_0  dissimilarity_0  homogeneity_0  energy_0  correlation_0  \
0  1952.343750         7.656250       0.970093  0.866311       0.712865   
1  2828.906250        11.093750       0.956665  0.816831       0.692669   
2  2346.796875         9.203125       0.964050  0.822255       0.747541   
3  1988.203125         7.796875       0.969543  0.866829       0.704665   
4  3882.773438        15.226562       0.940521  0.703014       0.752067   

      entropy_0     asm_0  contrast_45  dissimilarity_45  homogeneity_45  ...  \
0 -5.105249e+06  0.000013       3084.0         12.094118        0.952757  ...   
1 -7.070763e+06  0.000012       4408.0         17.286275        0.932475  ...   
2 -7.147066e+06  0.000013       3870.0         15.176471        0.940717  ...   
3 -5.051554e+06  0.000013       3256.0         12.768627        0.950123  ...   
4 -1.291644e+07  0.000011       8468.0         33.207843        0.870282  ...   

     entropy_95    asm_90  contrast_135  dissimilarity_135  homogeneity_135  \
0 -5.105249e+06  0.000013        3044.0          11.937255         0.953370   
1 -7.070763e+06  0.000013        4596.0          18.023529         0.929596   
2 -7.147066e+06  0.000012        4188.0          16.423529         0.935846   
3 -5.051554e+06  0.000013        3220.0          12.627451         0.950674   
4 -1.291644e+07  0.000010        8506.0          33.356863         0.869700   

   energy_135  correlation_135   entropy_135   asm_135  labels  
0    0.850427         0.553960 -5.105249e+06  0.000013       0  
1    0.792249         0.502481 -7.070763e+06  0.000012       0  
2    0.796272         0.551083 -7.147066e+06  0.000012       0  
3    0.849020         0.523450 -5.051554e+06  0.000013       0  
4    0.644671         0.458630 -1.291644e+07  0.000010       0  

[5 rows x 29 columns]

In [55]:
len(clean_data.columns)

29

In [56]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix


x = clean_data.iloc[:,:-1]
y = clean_data.iloc[:,-1:]

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=0)

scaler = MinMaxScaler()
x_train_scalled = scaler.fit_transform(x_train)
x_test_scalled = scaler.transform(x_test)

##### Cek dokumentasi parameter pemanggilan fungsi pada list classifiers

In [57]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
names = ["K-Nearest Neighbors", 
         "Linear SVM"]

classifiers = [
    KNeighborsClassifier(5),
    SVC(gamma=2, C=1),
    ]

#### Lakukanlah analisis hasil prediksi

In [58]:
for name, model in zip(names, classifiers):
  model.fit(x_train_scalled, y_train.values.ravel())
  y_pred_model = model.predict(x_test_scalled)
  print(f'## MODEL NAME : {name}')
  print(confusion_matrix(y_test, y_pred_model))
  print(classification_report(y_test, y_pred_model))
  print('\n')

## MODEL NAME : K-Nearest Neighbors
[[14  4  2]
 [ 3 16  1]
 [ 0  2 18]]
              precision    recall  f1-score   support

           0       0.82      0.70      0.76        20
           1       0.73      0.80      0.76        20
           2       0.86      0.90      0.88        20

    accuracy                           0.80        60
   macro avg       0.80      0.80      0.80        60
weighted avg       0.80      0.80      0.80        60



## MODEL NAME : Linear SVM
[[16  4  0]
 [ 4 16  0]
 [ 0  4 16]]
              precision    recall  f1-score   support

           0       0.80      0.80      0.80        20
           1       0.67      0.80      0.73        20
           2       1.00      0.80      0.89        20

    accuracy                           0.80        60
   macro avg       0.82      0.80      0.81        60
weighted avg       0.82      0.80      0.81        60





In [59]:
correlation = {}
label = clean_data['labels']
for col in clean_data.columns:
  correlation[col] = label.corr(clean_data[col])
sorted_correlation = sorted(correlation.items(), key=lambda x: x[1])
print(sorted_correlation)
len(sorted_correlation)

[('energy_0', -0.650217936307868), ('asm_0', -0.650217936307868), ('asm_45', -0.6401500005948308), ('energy_45', -0.6401500005948307), ('energy_135', -0.6392842999348268), ('asm_135', -0.6392842999348268), ('homogeneity_0', -0.6355944380968352), ('entropy_0', -0.623849231809006), ('entropy_45', -0.623849231809006), ('entropy_95', -0.623849231809006), ('entropy_135', -0.623849231809006), ('asm_90', -0.6231862571884728), ('energy_90', -0.6231862571884726), ('homogeneity_135', -0.6018139494478015), ('homogeneity_45', -0.6012720159601499), ('homogeneity_90', -0.5181705582033235), ('correlation_0', -0.4844778045307333), ('correlation_45', -0.21351914346184683), ('correlation_135', -0.20140361880631066), ('correlation_90', -0.03523691597837949), ('contrast_90', 0.5181705582033233), ('dissimilarity_90', 0.5181705582033237), ('contrast_45', 0.6012720159601496), ('dissimilarity_45', 0.6012720159601496), ('contrast_135', 0.6018139494478019), ('dissimilarity_135', 0.601813949447802), ('dissimilar

29

In [60]:
positive_corr = {}
negative_corr = {}
for (k,v) in sorted_correlation:
  if(v>0):
    positive_corr[k] = v
  else:
    negative_corr[k] = v
len(positive_corr)

9

### PERBANDINGAN
Dalam membandingkan kedua model, K-Nearest Neighbors (KNN) dan Linear SVM, terdapat perbedaan sebagai berikut:

Akurasi: Akurasi berfungsi untuk mengukur sejauh mana model dapat memprediksi dengan benar. Pada model KNN, akurasi sebesar 0.60, sedangkan pada model Linear SVM, akurasi mencapai 0.73. Oleh karena itu, Linear SVM memiliki akurasi yang lebih tinggi daripada KNN.

Presisi dan Recall: Presisi mengukur sejauh mana hasil positif yang diprediksi oleh model adalah benar, sedangkan recall mengukur sejauh mana model dapat mengidentifikasi secara lengkap hasil positif yang sebenarnya. Model KNN memiliki presisi dan recall yang lebih rendah dibandingkan dengan model Linear SVM.

F1-Score: F1-score merupakan rata-rata antara presisi dan recall. Baik KNN maupun Linear SVM memiliki f1-score yang lumayan tinggi, tetapi model Linear SVM memiliki nilai f1-score yang sedikit lebih tinggi daripada KNN.


### KESIMPULAN
Berdasarkan perbandingan kedua model tersebut, dapat disimpulkan bahwa model Linear SVM memiliki hasil yang lebih baik daripada model K-Nearest Neighbors dalam memprediksi kelas pada dataset yang diberikan. Model Linear SVM memiliki akurasi yang lebih tinggi dan secara konsisten memberikan nilai presisi, recall, dan f1-score yang lebih baik untuk setiap kelas. Oleh karena itu, model Linear SVM direkomendasikan untuk digunakan dalam klasifikasi project ini.